source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -293,6 +310,44 @@ void SslSocket::onConnected() { ASSERT(info_->state() == Ssl::SocketState::PreHa
 
 Ssl::ConnectionInfoConstSharedPtr SslSocket::ssl() const { return info_; }
 
+void SslSocket::keylogCallback(const SSL* ssl, const char* line) {
+  ASSERT(ssl != nullptr);
+  BIO* bio_keylog = static_cast<BIO*>(SSL_get_ex_data(ssl, ssl_ex_data_index_));
+  BIO_printf(bio_keylog, "%s\n", line);
+  (void)BIO_flush(bio_keylog);
+}
+
+void SslSocket::disableTlsKeyLog(void) {
+  if (bio_keylog_ != nullptr) {
+    BIO_free_all(bio_keylog_);
+    bio_keylog_ = nullptr;
+  }
+  SSL_CTX* ctx = SSL_get_SSL_CTX(rawSsl());
+  ASSERT(ctx != nullptr);
+  SSL_CTX_set_keylog_callback(ctx, nullptr);
+  enable_tls_keylog_ = false;
+}
+
+void SslSocket::enableTlsKeyLog(void) {
+  SSL_CTX* ctx = SSL_get_SSL_CTX(rawSsl());
+  ASSERT(ctx != nullptr);
+  /*
+   * Append rather than write in order to allow concurrent modification.
+   * Furthermore, this preserves existing key log files which is useful when
+   * the tool is run multiple times.
+   */
+  bio_keylog_ = BIO_new_file(config_->getTlsKeyLogPath().c_str(), "a");
+  if (bio_keylog_ == nullptr) {
+    ENVOY_LOG(warn, "Creating BIO file fails, log path: {}", config_->getTlsKeyLogPath().c_str());
+    return;
+  }
+
+  SSL* ssl = rawSsl();
+  SSL_set_ex_data(ssl, SslSocket::ssl_ex_data_index_, bio_keylog_);
+  SSL_CTX_set_keylog_callback(ctx, keylogCallback);
```

Comment:
This is setting a function on the SSL_CTX, which is shared among all connections, which isn't correct from ssl_socket. 

I think you should restructure to do this in `ContextImpl`, and do the filtering for whether to log a given key in the `keylogCallback` function. It will get called for every connection, and you'll only log if that connection matches the configured criteria.

Reply:
@ggreenway , if we use access log, it seems that we must configure the access logger in TLS and create a new access logger for TLS. Currently there is no access logger for transport socket.

Reply:
You can add it to the TransportSocketFactoryContext and get it there.

Reply:
fixed